specs/agents/tracing-spans-service-target.md

```diff
@@ -0,0 +1,236 @@
+
+    } else if (context.http.url) { // http spans
+      service_target.name = context.http.host;
+      if (context.http.url.port > 0) {
+        service_target.name += ":" + context.http.port;
+      }
+    }
+
+  }
+} else {
+    // non-exit spans should not have service.target.* fields
+    span.context.service.target = undefined;
+}
+```
+
+The values for `span.context.db.instance` are described in [SQL Databases](./tracing-instrumentation-db.md#sql-databases).
+
+The values for `span.context.message.queue.name` are described in [Messaging context fields](./tracing-instrumentation-messaging.md#context-fields)
+
+### User API
+
+Agents SHOULD provide an API entrypoint to set the value of `span.context.destination.service.resource`,
+setting an empty or `null` value allows the user to discard the inferred value.
+This API entrypoint should be marked as deprecated and replaced by the following:
+
+Agents SHOULD provide an API entrypoint to set the value of `span.context.service.target.type` and `span.context.service.target.name`,
+setting an empty or `null` value on both of those fields allows the user to discard the inferred values.
+
+When a user-provided value is set, it should take precedence over inferred values from the span `_.type` `_.subtype` or any `_.context` attribute.
+
+### Phase 3
+
+In Phase 3 the `span.destination.service.resource` field value is inferred on APM Server with the following algorithm:
+
+```javascript
+
+// span from agent intake
+span = {};
+
+// inferred destination resource (if any)
+destination_resource = undefined;
+
+if (span.service.target.type) {
+  // new fields, store them as-provided
+  // infer resource destination from new fields
+  destination_resource = span.service.target.type;
+  if (span.service.target.name) {
+    destination_resource += "/";
+    destination_resource += span.service.target.name;
+  }
+
+} else if (['db', 'messaging', 'external'].indexOf(span.type) >= 0) {
+  // agent intake without new fields
+  // only a subset of span types should be included
+  span.service.target = {};
+
+  // inferred service target type & name
+  target_type = undefined;
+  target_name = undefined;
+
+  if (span.destination.service.resource) {
+    // infer new fields values from destination resource if provided
+    r = span.destination.service.resource;
+    separatorIndex = r.indexOf('/');
+    if (separatorIndex <= 0) {
+      target_type = r;
+    } else {
+      target_type = r.substr(0, separatorIndex);
+      target_name = r.substr(separatorIndex + 1);
+    }
+  } else {
+    // no destination resource, fallback from span type & subtype
+    target_type = span.type;
+    if (span.subtype) {
+      target_name = span.subtype;
+    }
+  }
+
+  if (target_name.length == 0) {
+    // normalization: do not store empty name
+    target_name = undefined;
+  }
+
+  span.service.target.type = target_type;
+  span.service.target.name = target_name;
+
+}
```

Comment:
@lahsivjar @AlexanderWert and I were discussing this part and wondered whether we can simplify to: "If the new field (`span.service.target.type`) is provided, backfill old field (`span.destination.service.resource`). Otherwise, if the old field is set, infer the new fields."

But maybe we were missing a nuance why there's another fallback for specific `span.type`s. Can't we just ignore cases where both `span.service.target.type` and `span.destination.service.resource` is not set?

```suggestion
} else if (span.destination.service.resource) {
  // infer new fields values from destination resource if provided
  r = span.destination.service.resource;
  if (r.contains('/')) {
    span.service.target.type = r.substr(0, r.indexOf('/'));
    span.service.target.name = r.substr(r.indexOf('/') + 1);
  } else {
    span.service.target.type = r;
  }
}
```

Reply:
The conditional is only here to limit the scope of the impacted services, I would be fine to remove those but it means that we also have to specify the `service.target.{type,name}` for all the possible values of `span.subtype`. For example the Azure and AWS instrumentations might already define specific values for destination resource field.

------

There is a corner case where it might sense for the agent to send both: when the `host:port` format is currently used for resource, for example with an HTTP client request:
- `span.destination.service.resource` = `host:80` 
- `span.service.target.type` = `http`
- `span.service.target.name` = `host:80`

In this case, the agent already has some logic to deal with that, so we probably don't want to add another mapping on the server to have rules like "if `service.target.type` = `http` then only take `service.target.name` to infer the `span.destination.service.resource` value.

-----

So I think here the best approach would be to deal with 3 cases:
- both old and new fields are provided: use them as-provided (Phase1 for server + Phase 2 for agent(s))
- old fields are provided but not the new ones: infer the new ones (Phase 1 for server)
- new fields are provided but not the old ones: infer the old ones (Phase 3 for server + Phase 4 & 5 for agent(s))

Reply:
After thinking a bit more on this, my previous answer :top: is probably not the best solution as it makes it harder for agents to completely drop the `resource` field.

Inferring the new/old fields can be simplified to:
- if the new fields are present, then infer `resource`, ignore provided `resource` if there is any (1)
- else, if the old fields are present, try to infer the new ones (2)

(1) infer from new fields:
- when span type is `external` or `storage`, set `resource` = `${service.target.name}` (with fallback on `${service.target.type}`), this will cover the case of HTTP spans previously mentioned
- when span type is NOT `external` or `storage`, set `resource` = `${service.target.type}/${service.target.name}` (with fallback on `${service.target.type}`, this will cover the general case of DB spans.

(2) infer new fields from old ones:
- when span type is `db` or `messaging`, the `resource` should be in the `${service.target.type}` or `${service.target.type}/${service.target.name}` format, so try to parse `type` and `name` from this.
- when span type is NOT in `db` or `messaging`, set `service.target.type` to `${span.subtype}` and `service.target.name` to `${resource}` (thus without trying to parse it).

@felixbarny @lahsivjar @AlexanderWert what do you think of this simpler approach ?

Reply:
Doing different things based on different types sounds brittle and complex. 
Is there a way we can do the same simple strategy consistently? I'd be willing to make tradeoffs for the case where the old field has to be back-filled.

Also, do agents need to keep sending both the old and the new fields? If so, why are we back-filling the old field in APM Server?

Reply:
Agents have to stay compatible with apm-server < 8.3, so they MUST compute the `destination.service.resource` anyway, even if they MUST NOT rely on it internally (dropped spans, compressed spans, ...) : I need to update the spec to reflect that.

We could imagine a future version (yet undefined) of the stack where agents do not send this field anymore because it's not used by UI anymore, but this would likely be a breaking change and unlikely to be short/medium-term.

Inferring the `destination.service.resource` value from the new `service.target.{type,name}` fields on APM server would require to duplicate parts of the agents logic to apm-server, with the small benefit of allowing agents not to send the resource, but they still need to have it for older apm-server versions.

As a consequence, maybe the best compromise for now would be to make the agents keep sending the `resource` field alongside the new fields. We will only be able to get rid of it once the UI does not rely on it anymore, and the new fields (either from the agent or inferred by server) allow to future-proof the captured data, which might help minimize the impact of a probable future breaking change in this area.

On the other end, for OTel spans intake, we still have to infer the equivalent values of `destination.service.resource` and the new fields `service.target.{type,name}` from the OTel attributes, so having this kind of logic on APM-server based on span type/subtype (or other attributes) would just be an extra mapping, and we could maybe make the APM server reuse the JSON specs we use for agents for span types/subtypes & resource destination.

I think for now the best option is to keep sending the `destination.service.resource` as it's the simplest thing we can do.
---

Another half-baked idea here: for the case of the HTTP spans, we could imagine adding an extra `service.target.address` and make apm-server generate the `destination.service.resource` from it when the service.target.name is not set,
but that would not work for the AWS S3 spans or custom user spans where only the name is currently used, also for spans like gRPC, having all of the service.target.{type,name,address} set makes sense at instrumentation but would produce an unexpected resource value.